src/base/utils/version.h

```diff
@@ -0,0 +1,149 @@
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef QBITTORRENT_UTILS_VERSION_H
+#define QBITTORRENT_UTILS_VERSION_H
+
+#include <array>
+#include <stdexcept>
+
+#include <QString>
+#include <QStringList>
+
+template <class T, std::size_t N>
+class Version
+{
+public:
+    typedef T ComponentType;
+    typedef Version<T,N> ThisType;
+
+    Version()
+    {
+        m_components.fill(T());
+    }
+
+    template <class ... Other>
+    Version(Other ... components)
+        : m_components {{components ...}}
+    {
+    }
+
+    Version(const QString &version)
+        : Version {version.split(QLatin1Char('.'))}
+    {
+    }
+
+    Version(const QByteArray &version)
+        : Version {version.split('.')}
+    {
+    }
+
+    ComponentType majorNumber() const
+    {
+        static_assert(N >= 1, "Array size is too small");
+        return (*this)[0];
+    }
+
+    ComponentType minorNumber() const
+    {
+        static_assert(N >= 2, "Array size is too small");
+        return (*this)[1];
+    }
+
+    ComponentType revisionNumber() const
+    {
+        static_assert(N >= 2, "Array size is too small");
+        return (*this)[2];
+    }
+
+    ComponentType patchNumber() const
+    {
+        static_assert(N >= 3, "Array size is too small");
+        return (*this)[3];
+    }
+
+    ComponentType operator[](std::size_t i) const
+    {
+        return m_components.at(i);
+    }
+
+    operator QString() const
+    {
+        QString res = QString::number((*this)[0]);
+        for (std::size_t i = 1; i < N; ++i)
+            res += QLatin1Char('.') + QString::number((*this)[i]);
+        return res;
+    }
+
+    bool operator==(const ThisType &other) const
+    {
+        return m_components == other.m_components;
+    }
+
+private:
+    template <class QtStringsList>
+    Version(const QtStringsList &versionParts)
+    {
+        if (versionParts.size() != N)
+            throw std::logic_error("Incorrect number of version components");
+        bool ok = false;
+        for (std::size_t i = 0; i < N; ++i) {
+            m_components[i] = static_cast<T>(versionParts[i].toInt(&ok));
+            if (!ok)
+                throw std::runtime_error("Can not parse version components");
+        }
+    }
+
+    std::array<T, N> m_components;
+};
+
+template <class T, std::size_t N>
+bool operator!=(const Version<T, N> &left, const Version<T, N> &right)
```

Comment:
It depends, as you probably know. For this class I prefer members: less typing, no profits from non-members.